vm/sdk/src/wallet.rs

```diff
@@ -0,0 +1,61 @@
+//! Definitions of the reference wallet interface.
+
+use athena_interface::Address;
+use parity_scale_codec::{Decode, Encode};
+
+use crate::Pubkey;
+
+#[derive(Clone, Copy, Debug, Default, Encode, Decode, PartialEq, Eq)]
+pub struct SpendArguments {
+  pub recipient: Address,
+  pub amount: u64,
+}
+
+pub trait WalletProgram {
+  fn spawn(owner: Pubkey) -> Address;
+  fn spend(&self, args: SpendArguments);
+  fn proxy(&self, destination: Address, args: &[u8]);
+  fn deploy(&self, code: Vec<u8>) -> Address;
+}
+
+pub fn encode_spend(state: Vec<u8>, args: SpendArguments) -> Vec<u8> {
```

Comment:
Why does encoding require the state? I thought encoding consisted of method + args (i.e., the payload type we defined)

Reply:
I see that the method selector is being applied at the higher level, okay, but I'm still confused why this requires state.

Reply:
Because the encoded state of the program is also one of the inputs and must be written to the "stdin" along with the actual arguments so that the program side can read it from the IO and decode. For example, look at the spend method definition:
```rust 
fn spend(&self, args: SpendArguments)
```

Both `self` and the `args` come from the "message input".

Reply:
Okay, but it's important to distinguish between the state and the payload, which consists of the method selector and args only. Maybe we need yet another layer here to encapsulate all of these things? It's important because the tx itself contains the payload, but it does not contain the state (which could be quite large). The host (go-sm) has to pass the state into the VM.

By the way, it turns out that spend() actually does not depend on the state:

https://github.com/athenavm/athena/blob/76e6130e26527a61cbf86b8a5950ed91273a69a5/examples/wallet/program/src/main.rs#L41-L46

so it could be turned into an "associated function" (one that doesn't take self) like spawn. It depends on _context_ but not on _state._ verify(), on the other hand, does depend on state.

Reply:
>Okay, but it's important to distinguish between the state and the payload, which consists of the method selector and args only. Maybe we need yet another layer here to encapsulate all of these things? It's important because the tx itself contains the payload, but it does not contain the state (which could be quite large). The host (go-sm) has to pass the state into the VM.

Right, the TX payload contains `method_selector | arguments` but cannot contain `state` so encoding it `athcon_encode_spend()` makes no sense. On the other hand, the `runner`  must pass the state to stdin in its `execute()` method:

```rust
let mut stdin = AthenaStdin::new();
// same as stdin.write_vec(state | args);
stdin.write_vec(state);
stdin.write_vec(args);
```

So the question is, how does the `runner` obtain the state? Should it call the `host` or assume that `msg.input_data` already contains it (i.e., `input_data = state | execution_payload(method | args)`)?

>By the way, it turns out that spend() actually does not depend on the state:

Then, what is the purpose of `spend()` beyond proxying to the `host::call()`? What logic could a contract implement there? I'm trying to understand why this method is even needed in the `WalletProgram`.

Reply:
> Then, what is the purpose of spend() beyond proxying to the host::call()? What logic could a contract implement there? I'm trying to understand why this method is even needed in the WalletProgram.

There's more complex logic for multisig: there can be spend limits per pubkey, daily spend limits, etc.

Reply:
Let's discuss the rest at https://github.com/athenavm/athena/pull/157#discussion_r1804309118